eip-0006.md

```diff
@@ -0,0 +1,259 @@
+# EIP-0006: Informal Multi-Stage Contract Specification Format
+
+* Author: Robert Kornacki
+* Status: Proposed
+* Created: 21-Apr-2020
+* License: CC0
+* Forking: No fork needed 
+
+### Contents
+- [Background And Motivation](#background-and-motivation)
+- [How Informal Multi-Stage Contract Specs Are Written](#How-Informal-Multi-Stage-Contract-Specs-Are-Written)
+- [Alternative Formatting](#Alternative-formatting)
+- [Example NFT Auction Contract](#example-nft-auction-contract)
+- [Conclusion](#Conclusion)
+
+### Background And Motivation
+
+[Multi-stage contracts](https://link.springer.com/chapter/10.1007/978-3-030-31500-9_16) are a fascinating tool for implementing complex protocols on a UTXO-based system. They allow for great expressiveness and are generally approachable due to the fact that different phases of a protocol are cleanly separated into their own sub-contracts.
+
+With that said, multi-stage contracts are extremely nascent as a model. This means that though they may be simpler when compared to alternatives, complexity is still going to be present when dealing with protocols that require numerous stages. Furthermore, we could benefit from conveying to others how a protocol is supposed to be implemented and/or work without writing a single line of code.
+
+As such it would be quite handy to have a format to informally define specifications for our multi-stage contracts that is understandable, easily traversable, and can be used as a guide for eventually writing both the on-chain and off-chain code.
+
+
+
+### How Informal Multi-Stage Contract Specs Are Written
+
+Let's look at the markdown-based schema in it's entirety and drill down how it works below.
+
+```md
+# Contract Name
+Contract preamble
+
+## ToC
+1. [Bootstrap Stage](<#Bootstrap-Stage>)
+2. [X Stage](<#X-Stage>)
+
+## Bootstrap Stage
+Preamble of how the contract is bootstrapped. 
+
+### List Of Paths
+- [Y Path](<#Y-Path>)
+ 
+### Y Path
+Preamble
+
+#### Inputs
+1. ...
+2. ...
+3. ...
+
+
+#### Outputs
+1. ...
+2. ...
+3. ...
+
+
+#### Leads To
+[X Stage](<#X-Stage>)
+
+---
+
+## X Stage
+Preamble
+
+### Hard-coded Values
+- ...
+
+### Registers
+- R4: ...
+- R5: ...
+
+
+### List Of Paths
+- [Y Path](<#Y-Path>)
+ 
+### Y Path
+Preamble
+
+#### Inputs
+1. ...
+2. ...
+3. ...
+
+
+#### Outputs
+1. ...
+2. ...
+3. ...
+
+
+#### Path Conditions
+1. ...
+2. ...
+
+
+#### Leads To
+`Hyperlinked name of next stage` OR `Exit`
+
+---
+
+```
+
+As you can see above, each contract is made out of a number of **stages**. Each stage (except for the bootstrap stage) is a specific contract that the locks up tokens and data. Every stage has the ability to specify values which are hard-coded within the contract, and/or values which are stored in the registers.
```

Comment:
Right, these informal specs don't need to map onto a single person's viewpoint as a separate FSM. The DEX would be in a single multi-stage contract spec and you would have a separate "Buyer Bootstrap Stage" And "Seller Bootstrap Stage" which I mention multiple bootstrap phases are possible later in the document.

In regards to Swap transaction, in the buyer's box/stage you'd have a path that is labeled "Swap Path" and in the sellers box/stage you'd have a "Swap Path" as well. In these paths you clarify that as inputs (and in the conditions) the other person's box is required.

Essentially you have this cross-referencing going on where each actor is in a different stage, and their stages can intersect and be spent together into a new stage.

Reply:
@robkorn It becomes too complicated IMO. 
The goal of this formalized spec is to make it clear to anyone, right? 
I think structuring the template after FSM is semantically problematic, because FSM is not the right model. It make sense to have stages though.

Reply:
The reason we are using markdown is to allow for linking in the ToC, the list of paths, and where each path leads to so you can follow from a single bootstrap stage -> all the way to the end of a protocol by following the paths. This way you don't get lost and you can understand exactly how one actor in the protocol must move forward.

I don't really understand how it's any more complicated to you than FlowCards for example. It may be verbose, but FlowCards are inherently a lot harder to understand because people need to understand the Ergo model of contracts, how state transitions work, the conditions which you write in a very short form, etc. Everything is laid out in English here, and it's considerably easier to understand because you have separate Bootstrap phases for each actor which you can follow along. FlowCards don't have clear entry-points for someone to start looking at where the protocol begins on the other hand.

This may end up just being difference in perspective/opinion, but FlowCards are genuinely much more complicated to me while this format is straight english what the protocol does with links to make it easier to traverse.